pilot/pkg/model/auth_cache.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	authn "istio.io/api/authentication/v1alpha1"
+	"istio.io/istio/pkg/cache"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	// https://openid.net/specs/openid-connect-discovery-1_0.html
+	// OpenID Providers supporting Discovery MUST make a JSON document available at the path
+	// formed by concatenating the string /.well-known/openid-configuration to the Issuer.
+	openIDDiscoveryCfgURLSuffix = "/.well-known/openid-configuration"
+
+	// OpenID Discovery web request timeout.
+	openIDDiscoveryHTTPTimeOutInSec = 5
+
+	// JwksURI Cache expiration time duration, individual cached JwksURI item will be removed
+	// from cache after its duration expires.
+	jwksURICacheExpiration = time.Hour * 24
+
+	// JwksURI Cache eviction time duration, cache eviction is done on a periodic basis,
+	// jwksURICacheEviction specifies the frequency at which eviction activities take place.
+	jwksURICacheEviction = time.Minute * 30
+)
+
+// jwksURIResolver is wrapper of expiringCache for jwksUri.
+type jwksURIResolver struct {
+	cache cache.ExpiringCache
+}
+
+// newJwksURIResolver creates new instance of jwksURIResolver.
+func newJwksURIResolver() jwksURIResolver {
+	return jwksURIResolver{cache.NewTTL(jwksURICacheExpiration, jwksURICacheEviction)}
+}
+
+// Set jwks_uri through openID discovery if it's not set in auth policy.
+func (r *jwksURIResolver) setAuthenticationPolicyJwksURIs(policy *authn.Policy) {
+	if policy == nil {
+		return
+	}
+
+	for _, method := range policy.Peers {
+		switch method.GetParams().(type) {
+		case *authn.PeerAuthenticationMethod_Jwt:
+			policyJwt := method.GetJwt()
+			if policyJwt.JwksUri == "" {
+				uri, err := r.resolveJwksURIUsingOpenID(policyJwt.Issuer)
+				if err != nil {
+					log.Warnf("Failed to get jwks_uri for issuer %q: %v", policyJwt.Issuer, err)
```

Comment:
We should return error and may be reject policy. There are some assumption down the line that jwksURI must exist. Policy with invalid jwksURI will not work anyway.

Reply:
1. do we want to reject policy even if part of a list of jwksUri could be resolved ?
2. if yes for 1, I prefer to do it in a separated PR to avoid cascading change in this one.

Reply:
I think so. We used to assume policy is guarded by validation to prevent bad one; i.e they should be rejected even before getting to CRD. So it's may be ok to reject the policy here, though I have to say it's not a good the user experience (but let it goes through is equally bad). That's why I think we need to do something in the validation phase, or review what could be a better way to send feedback to users.

Reply:
updated to fallback to legacypolicy.